dns/resolver.py

```diff
@@ -1079,6 +1079,27 @@ def set_flags(self, flags):
 
         self.flags = flags
 
+    @property
+    def nameservers(self):
+        return self._nameservers
+
+    @nameservers.setter
+    def nameservers(self, nameservers):
+        """
+        :param nameservers: can be a ``str``, ``list``, or None.
+        If it's a ``str``, it will converted to a list.
+        :raise ValueError: if `nameservers` is anything other than \
+        ``str``, ``list``, or None.
+        """
+        if isinstance(nameservers, str):
+            self._nameservers = [nameservers]
+        elif isinstance(nameservers, list):
+            self._nameservers = nameservers
+        elif nameservers is None:
+            self._nameservers = None
```

Comment:
I think this will break stuff, `Resolver.__init__()` calls `reset()` which always produces a list (possibly empty). What would be meaning of `None`? Why not use empty list and get rid of `Optional` type annotation?

Reply:
> I think this will break stuff, `Resolver.__init__()` calls `reset()` which always produces a list (possibly empty).

True, it does produce an empty list, which is perfectly fine. If you assign it an empty list, `self._nameservers` will get assigned to an empty list in lines 1096-1097:

```python
elif isinstance(nameservers, list):
        self._nameservers = nameservers
```
 I'll add an empty list assignment to the test cases to make it more clear.

>What would be meaning of `None`?
>Why not use empty list and get rid of `Optional` type annotation?

I included `None` because `Resolver.__init__()` sets all of its attributes to `None` before calling `reset()`. I don't feel strongly either way if you'd rather take that part out (i.e. assign it to empty list instead of `None` before calling `reset()`.

I don't really know what `Optional` type annotation you're referring to. The changes I made to type hints (link - https://github.com/rthalley/dnspython/pull/406/files#diff-88bcd2907ddf9508e5d24d7bd67d0054R36) aren't related to the other changes I made. Honestly I probably should've just put that one-line change in a separate PR.

Reply:
Setting the internal `_nameservers` value to `None` seems needlesly risky. Consider the following scenario:

```
r.nameservers = None
r.nameservers.append(...)
```

I think the internal `Resolver._nameservers` value should be a list. There seems to be many places in the code that assume it is a list. It'd also be possible to add a mypy notation for the getter to indicate a list.

Setting `Resolver.nameserver` value to `None` could assign a new empty list to the internal variable.

Reply:
I agree with your comment (https://github.com/rthalley/dnspython/issues/403#issuecomment-571235029). It definitely adds needless complexity to allow anything other than a list.